### ISSUE TITLE
tf.layers.batch_normalization does not support fused

### ISSUE THREAD

**Issue by taion** [tensorflow/tensorflow#7549]
`tf.layers.batch_normalization` does not accept a `fused` argument, and appears to always use `tf.nn.batch_normalization` instead of `tf.nn.fused_batch_norm`. As such, it does not appear to be possible to follow performance best-practices when using `tf.layers.batch_normalization`.

### What related GitHub issues or StackOverflow threads have you found by searching the web for your problem?

Searched issues for "fused" and "batch_normalization"

### Environment info
Operating System: gcr.io/tensorflow/tensorflow:1.0.0-devel-gpu

### What other attempted solutions have you tried?
Currently shimming `tf.contrib.layers.batch_norm`

**Comment by tfboyd** [comment]
As you noticed fused has not been added to core.  Using the version of `layers.batch_norm` in contrib is the correct approach currently.  The feature will be added to core.  Leaving issue open until it is added.  I am excited that you looked at the guide.  I/We will be adding more "tips" in the near future.  Thank you for the feedback.

**Comment by tfboyd** [comment]
Small update.  I am still following up on this and more so now that I have a little more free time.

**Comment by Egor-Krivov** [comment]
I am also really looking forward to this to resolve. I am working with 5d tensors for 3d image analysis and fused batch norm from slim doesn't support 5d input (https://github.com/tensorflow/tensorflow/issues/5694) and is outdated. Right now I have to use my own function to implement something as basic as reasonably fast batch norm for 3d images. I have implemented the same network with fused batch norm in pytorch and with batch norm from tf.layers and it's about 15 times slower in training (I am using GPU). My function currently performs reshape to 4d and uses batch norm from slim and provides about the same performance as pytorch.

It is really surprising for me that this issue is so old. I would say that without fused batch norm current implementation should never be used, because of the performance (15 times slower in my case).

**Comment by taion** [comment]
@Egor-Krivov BTW, see https://github.com/tensorflow/tensorflow/issues/7551 – the massive perf issue arises whenever non-fused BN is applied to any axis other than the last.

**Comment by tfboyd** [comment]
This is in progress.  A an internal PR to add fused to tf.layers.batch_nornmalization is in progress.  We made it an top priority.  We do not promise dates because things can go wrong or issues come up but I expect it will be in before the end of the quarter.  Again, things can go wrong and that is not a promise just guidance.

**Comment by Egor-Krivov** [comment]
@taion Thank you for the remark. I was, indeed, using NCHWD format for my data, following performance guide, and probably got massive performance drop.
@tfboyd Thank you for fast response. Hope it will go as planned.
There really should be a big warning sign somewhere near tf.layers.batch_normalization, warning about https://github.com/tensorflow/tensorflow/issues/7551.

**Comment by taion** [comment]
@tfboyd 

It'd be good to add a note to the performance guide warning against using un-fused batch norm when not using a channels-last data format?

While both are active recommendations, it's really weird and strange that switching to NCHW without switching to fused batch norm gives such a massive slowdown – especially because "use NCHW" comes before "use fused batch norm" in the performance guide.

**Comment by tfboyd** [comment]
@taion  Thanks Jimmy, I will add that.  So much to do and not enough time in the day.  :-)  I just rewrote most of the code that the Hong Kong University uses for their tests and started using the [DataSets](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/contrib/data) which is new in TF 1.2 (RC right now).  I have found it to be much easier and doing side by side testing with queues it is significantly faster.  For something like CIFAR-10 and your code (I think you were using the dict approach) it is really easy to switch.  I have some example code that is not super clean but it covers the basics.  

Also, thank you for all of the thoughts for the perf guide.  My lack of updates is not due to not wanting to do them.  As I personally dig deeper into the examples and rewrite some code I am running into similar issues as yourself.  I hope this helps me advocate better for the performance changes I and "we" (the community) want.
